File: modulo04-agentes-autonomos/agent-evals-java/src/main/java/com/iadeva/evals/evaluator/ToolSelectionEvaluator.java
package com.iadeva.evals.evaluator;

// Equivalente ao tool_selection_eval.py da aula 12 — avalia qualidade das decisões de tool selection

import com.iadeva.evals.model.ToolSelectionCase;
import com.iadeva.evals.model.ToolSelectionReport;
import org.springframework.ai.chat.client.ChatClient;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.util.List;
import java.util.Map;
import java.util.regex.Pattern;

/**
 * Avalia a qualidade das decisões de seleção de ferramentas do agente.
 * Usa o ChatClient do Spring AI para simular decisões reais do modelo.
 * O agente recebe um contexto e precisa escolher a ferramenta correta.
 */
@Component
public class ToolSelectionEvaluator {

    private final ChatClient chatClient;
    private final double minAccuracy;
    private final double maxUnnecessary;

    // Ferramentas disponíveis no domínio de monitoramento de incidentes
    private static final List<String> AVAILABLE_TOOLS = List.of(
            "getMetrics", "getLogs", "getDeployHistory", "saveIncident", "notifyTeam"
    );

    public ToolSelectionEvaluator(
            ChatClient.Builder chatClientBuilder,
            @Value("${evals.tool-selection.min-accuracy:0.80}") double minAccuracy,
            @Value("${evals.tool-selection.max-unnecessary:0.10}") double maxUnnecessary
    ) {
        this.chatClient = chatClientBuilder.build();
        this.minAccuracy = minAccuracy;
        this.maxUnnecessary = maxUnnecessary;
    }

    /**
     * Executa a avaliação de tool selection sobre os 5 cenários hardcoded.
     * Retorna o relatório com métricas de acurácia e taxa de chamadas desnecessárias.
     */
    public ToolSelectionReport evaluate() {
        List<ToolSelectionCase> cases = buildTestCases();

        int correctTool = 0;
        int correctArgs = 0;
        int unnecessaryCalls = 0;
        int wrongTool = 0;

        for (ToolSelectionCase tc : cases) {
            String response = askModel(tc);
            String selected = extractToolName(response);

            boolean toolCorreto = tc.expectedTool().equalsIgnoreCase(selected);
            boolean ferramenta_proibida = tc.forbiddenTools().stream()
                    .anyMatch(f -> f.equalsIgnoreCase(selected));

            if (toolCorreto) {
                correctTool++;
                // Verifica se os argumentos-chave estão presentes na resposta
                boolean argsOk = tc.expectedArgs().keySet().stream()
                        .allMatch(k -> response.toLowerCase().contains(k.toLowerCase()));
                if (argsOk) correctArgs++;
            } else {
                wrongTool++;
            }

            if (ferramenta_proibida) {
                unnecessaryCalls++;
            }
        }

        int total = cases.size();
        double toolAccuracy = (double) correctTool / total;
        double argAccuracy = (double) correctArgs / total;
        double unnecessaryRate = (double) unnecessaryCalls / total;
        double wrongRate = (double) wrongTool / total;

        // Critério de aprovação: accuracy >= minAccuracy E unnecessaryRate <= maxUnnecessary
        boolean passed = toolAccuracy >= minAccuracy && unnecessaryRate <= maxUnnecessary;

        return new ToolSelectionReport(
                total,
                round(toolAccuracy),
                round(argAccuracy),
                round(unnecessaryRate),
                round(wrongRate),
                passed
        );
    }

    /**
     * Consulta o modelo para decidir qual ferramenta usar dado o contexto do caso de teste.
     */
    private String askModel(ToolSelectionCase tc) {
        String prompt = """
                Você é um agente de monitoramento de sistemas. Dado o contexto abaixo, escolha UMA ferramenta.
                
                Ferramentas disponíveis: %s
                
                Contexto: %s
                Passo do loop: %d
                
                Responda APENAS com o nome da ferramenta e os argumentos em formato JSON.
                Exemplo: {"tool": "getMetrics", "args": {"service": "payment-api"}}
                """.formatted(AVAILABLE_TOOLS, tc.context(), tc.loopStep());

        try {
            return chatClient.prompt()
                    .user(prompt)
                    .call()
                    .content();
        } catch (Exception e) {
            // Em caso de falha na API, retorna resposta vazia para contabilizar como errada
            System.err.println("[ToolSelectionEvaluator] Erro na chamada ao modelo: " + e.getMessage());
            return "{}";
        }
    }

    /**
     * Extrai o nome da ferramenta da resposta do modelo via regex.
     */
    private String extractToolName(String response) {
        Pattern pattern = Pattern.compile("\"tool\"\\s*:\\s*\"([^\"]+)\"");
        var matcher = pattern.matcher(response);
        return matcher.find() ? matcher.group(1) : "unknown";
    }

    /**
     * Define os 5 casos de teste de tool selection.
     * Cada caso representa uma situação real de um loop de agente de monitoramento.
     */
    private List<ToolSelectionCase> buildTestCases() {
        return List.of(
                new ToolSelectionCase(
                        "ts-001",
                        "Latência alta detectada em payment-api. Ainda não coletamos dados de métricas.",
                        1,
                        "getMetrics",
                        Map.of("service", "payment-api", "metric", "latency"),
                        List.of("saveIncident", "notifyTeam")
                ),
                new ToolSelectionCase(
                        "ts-002",
                        "Métricas coletadas mostram CPU em 95%. Preciso entender o que aconteceu antes de abrir incidente.",
                        2,
                        "getLogs",
                        Map.of("service", "payment-api", "level", "ERROR"),
                        List.of("saveIncident", "notifyTeam")
                ),
                new ToolSelectionCase(
                        "ts-003",
                        "Logs mostram OutOfMemoryError às 14h. Houve algum deploy próximo desse horário?",
                        3,
                        "getDeployHistory",
                        Map.of("service", "payment-api", "since", "13:00"),
                        List.of("saveIncident", "notifyTeam")
                ),
                new ToolSelectionCase(
                        "ts-004",
                        "Deploy identificado às 13:45. Causa raiz confirmada: memory leak na versão 2.3.1. Devo registrar o incidente.",
                        4,
                        "saveIncident",
                        Map.of("service", "payment-api", "severity", "high"),
                        List.of("getMetrics", "getLogs")
                ),
                new ToolSelectionCase(
                        "ts-005",
                        "Incidente registrado. Time de on-call precisa ser notificado imediatamente.",
                        5,
                        "notifyTeam",
                        Map.of("channel", "on-call", "severity", "high"),
                        List.of("getMetrics", "getLogs", "getDeployHistory")
                )
        );
    }

    private double round(double value) {
        return Math.round(value * 10000.0) / 10000.0;
    }
}

File: modulo04-agentes-autonomos/agent-evals-go/evaluator/tool_selection_evaluator.go
package evaluator

import (
	"context"
	"fmt"
	"regexp"
	"strings"

	"agent-evals/model"
)

// ChatClient é a única operação de que ToolSelectionEvaluator depende.
type ChatClient interface {
	Chat(ctx context.Context, userPrompt string) (string, error)
}

var availableTools = []string{"getMetrics", "getLogs", "getDeployHistory", "saveIncident", "notifyTeam"}

var toolNamePattern = regexp.MustCompile(`"tool"\s*:\s*"([^"]+)"`)

const toolSelectionPromptTemplate = `Você é um agente de monitoramento de sistemas. Dado o contexto abaixo, escolha UMA ferramenta.

Ferramentas disponíveis: %s

Contexto: %s
Passo do loop: %d

Responda APENAS com o nome da ferramenta e os argumentos em formato JSON.
Exemplo: {"tool": "getMetrics", "args": {"service": "payment-api"}}
`

// ToolSelectionEvaluator avalia a qualidade das decisões de seleção de
// ferramentas do agente — equivalente a ToolSelectionEvaluator.java
// (aula 12). Usa o ChatClient para simular decisões reais do modelo.
type ToolSelectionEvaluator struct {
	Client         ChatClient
	MinAccuracy    float64
	MaxUnnecessary float64
}

// Evaluate executa a avaliação de tool selection sobre os 5 cenários
// hardcoded (mesmos da versão Java).
func (e *ToolSelectionEvaluator) Evaluate(ctx context.Context) model.ToolSelectionReport {
	cases := buildTestCases()

	var correctTool, correctArgs, unnecessaryCalls, wrongTool int

	for _, tc := range cases {
		response := e.askModel(ctx, tc)
		selected := extractToolName(response)

		toolCorrect := strings.EqualFold(tc.ExpectedTool, selected)
		forbidden := false
		for _, f := range tc.ForbiddenTools {
			if strings.EqualFold(f, selected) {
				forbidden = true
				break
			}
		}

		if toolCorrect {
			correctTool++
			argsOK := true
			lowerResponse := strings.ToLower(response)
			for key := range tc.ExpectedArgs {
				if !strings.Contains(lowerResponse, strings.ToLower(key)) {
					argsOK = false
					break
				}
			}
			if argsOK {
				correctArgs++
			}
		} else {
			wrongTool++
		}

		if forbidden {
			unnecessaryCalls++
		}
	}

	total := len(cases)
	toolAccuracy := float64(correctTool) / float64(total)
	argAccuracy := float64(correctArgs) / float64(total)
	unnecessaryRate := float64(unnecessaryCalls) / float64(total)
	wrongRate := float64(wrongTool) / float64(total)

	minAccuracy := e.minAccuracy()
	maxUnnecessary := e.maxUnnecessary()
	passed := toolAccuracy >= minAccuracy && unnecessaryRate <= maxUnnecessary

	return model.ToolSelectionReport{
		TotalCases:            total,
		ToolSelectionAccuracy: round(toolAccuracy),
		ArgumentAccuracy:      round(argAccuracy),
		UnnecessaryCallsRate:  round(unnecessaryRate),
		WrongToolRate:         round(wrongRate),
		Passed:                passed,
	}
}

func (e *ToolSelectionEvaluator) minAccuracy() float64 {
	if e.MinAccuracy == 0 {
		return 0.80
	}
	return e.MinAccuracy
}

func (e *ToolSelectionEvaluator) maxUnnecessary() float64 {
	if e.MaxUnnecessary == 0 {
		return 0.10
	}
	return e.MaxUnnecessary
}

// askModel consulta o modelo para decidir qual ferramenta usar. Em caso de
// falha na API, retorna "{}" para contabilizar como resposta errada — mesmo
// comportamento defensivo da versão Java.
func (e *ToolSelectionEvaluator) askModel(ctx context.Context, tc model.ToolSelectionCase) string {
	prompt := fmt.Sprintf(toolSelectionPromptTemplate, availableTools, tc.Context, tc.LoopStep)

	response, err := e.Client.Chat(ctx, prompt)
	if err != nil {
		return "{}"
	}
	return response
}

func extractToolName(response string) string {
	matches := toolNamePattern.FindStringSubmatch(response)
	if matches == nil {
		return "unknown"
	}
	return matches[1]
}

// buildTestCases define os 5 casos de teste de tool selection — cada caso
// representa uma situação real de um loop de agente de monitoramento.
func buildTestCases() []model.ToolSelectionCase {
	return []model.ToolSelectionCase{
		{
			ID: "ts-001", Context: "Latência alta detectada em payment-api. Ainda não coletamos dados de métricas.",
			LoopStep: 1, ExpectedTool: "getMetrics",
			ExpectedArgs:   map[string]any{"service": "payment-api", "metric": "latency"},
			ForbiddenTools: []string{"saveIncident", "notifyTeam"},
		},
		{
			ID: "ts-002", Context: "Métricas coletadas mostram CPU em 95%. Preciso entender o que aconteceu antes de abrir incidente.",
			LoopStep: 2, ExpectedTool: "getLogs",
			ExpectedArgs:   map[string]any{"service": "payment-api", "level": "ERROR"},
			ForbiddenTools: []string{"saveIncident", "notifyTeam"},
		},
		{
			ID: "ts-003", Context: "Logs mostram OutOfMemoryError às 14h. Houve algum deploy próximo desse horário?",
			LoopStep: 3, ExpectedTool: "getDeployHistory",
			ExpectedArgs:   map[string]any{"service": "payment-api", "since": "13:00"},
			ForbiddenTools: []string{"saveIncident", "notifyTeam"},
		},
		{
			ID: "ts-004", Context: "Deploy identificado às 13:45. Causa raiz confirmada: memory leak na versão 2.3.1. Devo registrar o incidente.",
			LoopStep: 4, ExpectedTool: "saveIncident",
			ExpectedArgs:   map[string]any{"service": "payment-api", "severity": "high"},
			ForbiddenTools: []string{"getMetrics", "getLogs"},
		},
		{
			ID: "ts-005", Context: "Incidente registrado. Time de on-call precisa ser notificado imediatamente.",
			LoopStep: 5, ExpectedTool: "notifyTeam",
			ExpectedArgs:   map[string]any{"channel": "on-call", "severity": "high"},
			ForbiddenTools: []string{"getMetrics", "getLogs", "getDeployHistory"},
		},
	}
}
